### PR TITLE
💍 Fill ring map tool

### PR DESCRIPTION
**Pull request by 3nids**

### CODE REVIEW

**Comment by 3nids** [comment]
[Archive.zip](https://github.com/opengisch/QField/files/4292452/Archive.zip)

File: src/qml/FeatureForm.qml [hunk 1]
```diff
@@ -253,6 +253,8 @@ Page {
 
         Loader {
           id: attributeEditorLoader
+          asynchronous: true
+          visible: status == Loader.Ready
```
Comment (3nids) [inline src/qml/FeatureForm.qml]: that's one thing which helped.
As far as I tested it it doesn't cause any issue.

File: src/qml/geometry_editors/FillRingToolBar.qml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  DigitizingToolbar {
+    id: drawPolygonToolbar
+    showConfirmButton: true
+
+    QuestionDialog{
+      id: questionDialog
+    }
+
+    Loader {
+      id: formPopupLoader
```
Comment (3nids) [inline src/qml/geometry_editors/FillRingToolBar.qml]: that's the other thing, using a loader here avoid loading the form on the toolbar loading but later on demand.

File: src/core/qfieldutils.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+QgsFeature QFieldUtils::initFeature(QgsVectorLayer *layer, QgsGeometry geometry)
+{
+  QgsFeature f(layer->fields());
+  f.setGeometry(geometry);
+  return f;
+}
+
+QgsGeometry QFieldUtils::polygonFromRubberband(RubberbandModel *rubberBandModel, const QgsCoordinateReferenceSystem &crs)
```
Comment (m-kuhn) [inline src/core/qfieldutils.cpp]: Can we move this to geometryutils and featureutils etc?
I don't like this idea of a dumping ground

File: src/qml/FeatureForm.qml (hunk 1 above)
Comment (m-kuhn) [reply src/qml/FeatureForm.qml]: woohoo

File: src/qml/geometry_editors/FillRingToolBar.qml (hunk 2 above)
Comment (m-kuhn) [reply src/qml/geometry_editors/FillRingToolBar.qml]: interesting

File: src/core/qfieldutils.cpp (hunk 3 above)
Comment (3nids) [reply src/core/qfieldutils.cpp]: done, I moved them to a subrepo

**Comment by 3nids** [comment]
it will need a new SDK https://github.com/opengisch/OSGeo4A/pull/59

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-0a32e8d05466ac57f2f70a75841c326fa40564af-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-0a32e8d05466ac57f2f70a75841c326fa40564af-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-0a32e8d05466ac57f2f70a75841c326fa40564af-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-0a32e8d05466ac57f2f70a75841c326fa40564af-x86_64.apk)

**Comment by 3nids** [comment]
one the phone, I get some crashes (not always)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-af0450cc6885a198390dd7679e26d198c9cec538-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-af0450cc6885a198390dd7679e26d198c9cec538-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-af0450cc6885a198390dd7679e26d198c9cec538-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-af0450cc6885a198390dd7679e26d198c9cec538-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-a5837d595badd5c0de1c6bc74668e03cdd2a70d7-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-a5837d595badd5c0de1c6bc74668e03cdd2a70d7-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-a5837d595badd5c0de1c6bc74668e03cdd2a70d7-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-a5837d595badd5c0de1c6bc74668e03cdd2a70d7-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-712e87a8e1cf5b2b6a663a887e203ae5b4f1ab87-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-712e87a8e1cf5b2b6a663a887e203ae5b4f1ab87-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-712e87a8e1cf5b2b6a663a887e203ae5b4f1ab87-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-712e87a8e1cf5b2b6a663a887e203ae5b4f1ab87-x86_64.apk)

**Comment by 3nids** [comment]
I can't make it crash when using adb logcat and not on my computer anymore.....

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-61fc235205a44faaa7c6077cd063510a9d40a325-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-61fc235205a44faaa7c6077cd063510a9d40a325-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-61fc235205a44faaa7c6077cd063510a9d40a325-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-61fc235205a44faaa7c6077cd063510a9d40a325-x86_64.apk)

**Comment by 3nids** [comment]
Finally have the trace:

```
03-06 12:48:29.320 32492 32521 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x3 in tid 32521 (qtMainLoopThrea), pid 32492 (gis.qfield_beta)
03-06 12:48:29.472 32563 32563 I crash_dump32: obtaining output fd from tombstoned, type: kDebuggerdTombstone
03-06 12:48:29.472   998   998 I /system/bin/tombstoned: received crash request for pid 32521
03-06 12:48:29.474 32563 32563 I crash_dump32: performing dump of process 32492 (target tid = 32521)
03-06 12:48:29.497 32563 32563 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
03-06 12:48:29.498 32563 32563 F DEBUG   : Build fingerprint: 'xiaomi/jasmine/jasmine_sprout:10/QKQ1.190910.002/V11.0.5.0.QDIMIXM:user/release-keys'
03-06 12:48:29.498 32563 32563 F DEBUG   : Revision: '0'
03-06 12:48:29.498 32563 32563 F DEBUG   : ABI: 'arm'
03-06 12:48:29.498 32563 32563 F DEBUG   : Timestamp: 2020-03-06 12:48:29+0100
03-06 12:48:29.498 32563 32563 F DEBUG   : pid: 32492, tid: 32521, name: qtMainLoopThrea  >>> ch.opengis.qfield_beta <<<
03-06 12:48:29.498 32563 32563 F DEBUG   : uid: 10215
03-06 12:48:29.498 32563 32563 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x3
03-06 12:48:29.498 32563 32563 F DEBUG   : Cause: null pointer dereference
03-06 12:48:29.498 32563 32563 F DEBUG   :     r0  ffffffff  r1  ffffffff  r2  00000000  r3  e6683e87
03-06 12:48:29.498 32563 32563 F DEBUG   :     r4  ffffffff  r5  d9359480  r6  bc0d2f88  r7  c1d48d40
03-06 12:48:29.498 32563 32563 F DEBUG   :     r8  b4556940  r9  00000000  r10 d9359480  r11 ea29f260
03-06 12:48:29.498 32563 32563 F DEBUG   :     ip  bd9b2e78  sp  c1d48d38  lr  c0ccd8bb  pc  c0ccd900
03-06 12:48:29.499 32563 32563 F DEBUG   : 
03-06 12:48:29.499 32563 32563 F DEBUG   : backtrace:
03-06 12:48:29.499 32563 32563 F DEBUG   :       #00 pc 00081900  /data/app/ch.opengis.qfield_beta-BnqlFKakIW8jMNuxhDHPEQ==/lib/arm/libQt5Qml.so
``` 

not very helpful

**Comment by m-kuhn** [comment]
Crash:

- Happens on a property of FeatureModel

Trace:

image

Possibly related:

- https://codereview.qt-project.org/c/qt/qtdeclarative/+/223430/1
- https://codereview.qt-project.org/c/qt/qtdeclarative/+/228466/1
- https://bugreports.qt.io/browse/QTBUG-67118 (Specific fix for a similar / the same? issue on StackView)

**Comment by 3nids** [comment]
@m-kuhn thanks. The issues you are referencing are solved. Do you have any strategy to propose?

**Comment by m-kuhn** [comment]
> Certain QML types (in this case the Repeater) might hold
objects assigned to from C++. If these objects don't have
a parent they would get garbage collected even though
they are clearly still in use by the engine.

Is there a C++ created object to which a parent could be assigned?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-81d7d57ff454d9e4188f212f1c3770af2547aa87-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-81d7d57ff454d9e4188f212f1c3770af2547aa87-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-81d7d57ff454d9e4188f212f1c3770af2547aa87-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-81d7d57ff454d9e4188f212f1c3770af2547aa87-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-4de58a09fcd211a01384a810cbbff15fc066c37b-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-4de58a09fcd211a01384a810cbbff15fc066c37b-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-4de58a09fcd211a01384a810cbbff15fc066c37b-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-4de58a09fcd211a01384a810cbbff15fc066c37b-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-d309e012cdb191071a2024ba1b9f90acafb3ee23-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-d309e012cdb191071a2024ba1b9f90acafb3ee23-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-d309e012cdb191071a2024ba1b9f90acafb3ee23-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-d309e012cdb191071a2024ba1b9f90acafb3ee23-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-217ecc7fe15ed34b5ffe200c76c6626f66c9a072-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-217ecc7fe15ed34b5ffe200c76c6626f66c9a072-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-217ecc7fe15ed34b5ffe200c76c6626f66c9a072-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-857-217ecc7fe15ed34b5ffe200c76c6626f66c9a072-x86_64.apk)

**Comment by signedav** [comment]
When I press the icon image-in-image, QField crashes. I tested it with the advanced bee farming demo project. I get this in the output:

**Comment by signedav** [comment]
Update: It's not always crashing. Only when I don't do anything before (Right after installing QField, open Project, select Feature, edit Geometry, select Vertextool, select Ringtool.

That's the last thing happening meanwhile. Not sure if even anything to do with it.
```
W QField  : qrc:/qml/FeatureForm.qml:201:15: Unable to assign [undefined] to QModelIndex
W QField  : qrc:/qml/VisibilityFadingRow.qml:5:3: Unable to assign [undefined] to bool
W QField  : qrc:/qml/VisibilityFadingRow.qml:5:3: Unable to assign [undefined] to bool
```

I tried to debug. But there is nothing that useful: Stucked somewhere in the Disassembler in `QV4::QObjectWrappers::getProperty(QV4::ExecutionEngine *, QObje...`

**Comment by signedav** [comment]
Update: Forget about the log output above. It's crashing after them...

**Comment by signedav** [comment]
It happens on loading the `FillRingToolBar.qml` source.
Good possible it's something with the `EmbeddeFearureForm` again.

**Comment by m-kuhn** [comment]
Shall we revert for the release?

**Comment by signedav** [comment]
I see three possibilities:
- Wait with the release and invest time to solve it
- Release in beta while investing time to solve it knowing about the crash already
- Revert it

**Comment by 3nids** [comment]
Here I have another crash, when you click twice in a row on the fill ring icon (on the second time).
Indeed the crash occurs in this `QV4::QObjectWrappers::getProperty(` and I also think it's related to the embedded form. I will work on this tomorrow, I would propose to discuss on Monday?

**Comment by m-kuhn** [comment]
@3nids are you fine with releasing 1.5 without fill ring? I'd like to ship it once it's stable. Unless there is an urgent need to have it in this release - and doing a possibly big investment to make it rock solid under pressure.

**Comment by signedav** [comment]
@3nids I don't know how the revert commit b96afde5dae3456a319ee0b5a39b3c4cedfe5de5 comes to this branch. I made a revert on a new branch actually.

**Comment by 3nids** [comment]
> @3nids are you fine with releasing 1.5 without fill ring? 

Yes but:
* there is a small fix to the split tool included
* if I can make it by tomorrow, it would be nice if it's included so we get wider testing

**Comment by m-kuhn** [comment]
Alright, let's wait for monday :+1:

**Comment by signedav** [comment]
Wait with the release till Monday?

**Pull request merged**